quantization/src/encoded_vectors_pq.rs

```diff
@@ -97,80 +98,102 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
-            storage_builder.push_vector_data(&encoded);
-            end_index = next(end_index);
+    /// Encode whole storage inside rayon context
+    /// This function should be called inside `rayon::scope`
+    fn encode_storage_rayon<'a: 'b, 'b>(
+        scope: &rayon::Scope<'b>,
+        data: impl Iterator<Item = &'a [f32]> + Clone + Send + 'b,
+        storage_builder: &'b mut (impl EncodedStorageBuilder<TStorage> + Send),
+        vector_division: &'b [Range<usize>],
+        centroids: &'b [Vec<f32>],
+        max_threads: usize,
+    ) -> Result<(), EncodingError> {
+        let storage_builder = Arc::new(Mutex::new(storage_builder));
+
+        // Synchronization between threads. Use conditional variable for
+        // each thread. Each condvar is blocked instead of first.
+        // While encoding, thread `N` after `storage_builder` usage blocks themself and 
+        // unblock thread `N+1`.
+        // In summary, access to `storage_builder` is ordered by `thread_index` below.
+        let condvars = (0..max_threads)
+            .map(|_| Arc::new(ConditionalVariable::default()))
+            .collect::<Vec<_>>();
+        condvars[0].notify(); // Allow first thread to use storage
+
+        for thread_index in 0..max_threads {
+            // Thread process vectors `N` that `(N + thread_index) % max_threads == 0`.
+            let data = data.clone().skip(thread_index);
+            let storage_builder = storage_builder.clone();
+            let condvar = condvars[thread_index].clone();
+            let next_condvar = condvars[(thread_index + 1) % max_threads].clone();
+
+            scope.spawn(move |_| {
+                let mut encoded_vector = Vec::new();
+                for vector in data.step_by(max_threads) {
+                    encoded_vector.clear();
+                    Self::encode_vector(vector, vector_division, centroids, &mut encoded_vector);
+                    // wait for permission from prev thread to use storage
+                    condvar.wait();
+                    // push encoded vector to storage
+                    storage_builder
+                        .lock()
+                        .unwrap()
+                        .push_vector_data(&encoded_vector);
+                    // Notify next thread to use storage
+                    next_condvar.notify();
+                }
+            });
         }
+        Ok(())
     }
 
+    /// Encode single vector from `&[f32]` into `&[u8]`.
+    /// This method divides `vector_data` into chunks, for each chunk 
+    /// finds nearest centroid and replace whole chunk by nearest centroid index.
+    /// 
+    /// # Arguments
+    /// * `vector_data` - Original vector data
+    /// * `vector_division` - Division of original vector into chunks
+    /// * `centroids` - Centroid positions (flattened by chunks; for similarity to vector data format)
+    /// * `encoded_vector` - Encoded result as a preallocated vector
```

Comment:
Maybe we should assert in this function that `encoded_vector` is empty?

```rust
debug_assert!(encoded_vector.is_empty());
```

Reply:
Disagreed. I move here `encoded_vector.clear()`, `len` does not matter, I construst encoding result as `Vec::with_capacity(vector_division.len())`